interfaces/tests/sync_int/sync_int.c

```diff
@@ -38,7 +38,7 @@ static int increment_on_odd_values(void* address)
         (void)interlocked_increment(ptr);
         wake_by_address_all(ptr);
     }
-    ThreadAPI_Exit(0);
```

Comment:
>ThreadAPI_Exit(0); [](start = 4, length = 18)

This is actually needed IIRC for pthreads, So it should stay?

Reply:
Not according to the documentation.

[https://linux.die.net/man/3/pthread_exit](https://linux.die.net/man/3/pthread_exit) say: (also man 3 pthread_exit says the same thing):

```
Performing a **return** from the start function of any thread other than the main thread results in an implicit call to pthread_exit(), using the function's return value as the thread's exit status.
```

Here's another problem with "not returning" and actually exiting through ExitThread/pthread_exit. 

We've taken those functions and nicely thrown them in ThreadAPI_Exit. Well - that is confusing to the compiler (I've send you an email about this:) in Windows Because ExitThread is marked as "DECLSPEC_NORETURN" - (true that). So ... ThreadAPI_Exit is not marked as noreturn. So theeeeen... soooooome versions of visual studio will propagate the noreturn, and then real "return" cannot follow ThreadAPI_Exit. And soooooom ooooother versions of Visual Studio will not propagate NORETURN so then the compiler will complain about a function without a return value.

Having "return 0" solves both of these cases on Windows, because there's no ExitThread called.

Also return 0 is perfectly fine in Linux as far as I can tell. 0 being the same as NULL cause obviously LINUX needs to return a void* from its threads. How do we even... put a void* <--->DWORD, cause sizeof void* might be 4,8... depends... 

Anyway - I will make an item in **that** bucket about this discovery.

---
In reply to: [497240061](https://github.com/Azure/azure-c-pal/pull/38#discussion_r497240061) [](ancestors = 497240061)

Reply:
Ok, makes sense.
I know it is confusing to the compiler because I hit it first :-). When I enabled LTCG we started having that problem. At the time I disabled the warning, but since return does an implicit pthread_exit of course we want to remove Thread_Exit.

Cheers!

---
In reply to: [497247378](https://github.com/Azure/azure-c-pal/pull/38#discussion_r497247378) [](ancestors = 497247378,497240061)

Reply:
As a matter of fact we should remove ThreadAPI_Exit completely. It only "confuzles" to quote my daughter.

---
In reply to: [497251256](https://github.com/Azure/azure-c-pal/pull/38#discussion_r497251256) [](ancestors = 497251256,497247378,497240061)

Reply:
Task created to remove it: [https://msazure.visualstudio.com/One/_backlogs/backlog/Azure%20Messaging%20Store/Backlog%20items/?workitem=8444719](https://msazure.visualstudio.com/One/_backlogs/backlog/Azure%20Messaging%20Store/Backlog%20items/?workitem=8444719) 

---
In reply to: [497251649](https://github.com/Azure/azure-c-pal/pull/38#discussion_r497251649) [](ancestors = 497251649,497251256,497247378,497240061)